bindings/aws/s3/s3.go

```diff
@@ -44,37 +54,90 @@ func (s *AWSS3) Init(metadata bindings.Metadata) error {
 	if err != nil {
 		return err
 	}
-	uploader, err := s.getClient(m)
+	uploader, downloader, svc, err := s.getClient(m)
 	if err != nil {
 		return err
 	}
 	s.metadata = m
 	s.uploader = uploader
+	s.downloader = downloader
+	s.svc = svc
 
 	return nil
 }
 
 func (s *AWSS3) Operations() []bindings.OperationKind {
-	return []bindings.OperationKind{bindings.CreateOperation}
+	return []bindings.OperationKind{bindings.CreateOperation, bindings.GetOperation, bindings.DeleteOperation, CopyOperation}
 }
 
 func (s *AWSS3) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	bucket := s.metadata.Bucket
+	data := []byte{}
 	key := ""
-	if val, ok := req.Metadata["key"]; ok && val != "" {
-		key = val
+	err := (error)(nil)
+	if (req.Operation == bindings.CreateOperation) {
+		key = s.getKey(req)
 	} else {
-		key = uuid.New().String()
-		s.logger.Debugf("key not found. generating key %s", key)
+		key, err = s.getReqParam(req, "key")
+		if (err != nil) {
+			return nil, err
+		}
 	}
 
-	r := bytes.NewReader(req.Data)
-	_, err := s.uploader.Upload(&s3manager.UploadInput{
-		Bucket: aws.String(s.metadata.Bucket),
-		Key:    aws.String(key),
-		Body:   r,
-	})
+	switch op := req.Operation; op {
+	case bindings.CreateOperation:
+		r := bytes.NewReader(req.Data)
+		_, err := s.uploader.Upload(&s3manager.UploadInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(key),
+			Body:   r,
+		})
+		if (err != nil) {
+			s.logger.Errorf("s3 CREATE operation exceptions: %s", err)
+		}
+
+	case bindings.GetOperation:
+		buff := &aws.WriteAtBuffer{}
+		_, err = s.downloader.Download(buff, &s3.GetObjectInput{
+			Bucket: aws.String(s.metadata.Bucket),
+			Key:    aws.String(key),
+		})
+		if (err == nil) {
+			data = buff.Bytes()
+		} else {
+			s.logger.Debugf("s3 GET operation exceptions: %s", err)
+		}
+		
+	case bindings.DeleteOperation:
+		_, err = s.svc.DeleteObject(&s3.DeleteObjectInput{
+			Bucket: aws.String(s.metadata.Bucket),
+			Key:    aws.String(key),
+		})
+		if (err != nil) {
+			s.logger.Errorf("s3 DELETE operation exceptions: %s", err)
+		}
+
+	case CopyOperation:
+		source, err := s.getReqParam(req, "source")
+		if (err != nil) {
+			return nil, err
+		}
+		_, err = s.svc.CopyObject(&s3.CopyObjectInput{
+
+			CopySource: aws.String(url.PathEscape(source)),
+			Bucket: aws.String(s.metadata.Bucket),
+			Key: aws.String(key),
```

Comment:
'key' attribute seems to be ambiguous. Someone might need to know that there is 'source' attribute to realize that 'key' means 'destination'. Please, use a more descriptive attribute here.